File: src/main/java/ch/unibas/dmi/dbis/vrem/database/codec/ExhibitionCodec.java
package ch.unibas.dmi.dbis.vrem.database.codec;

import ch.unibas.dmi.dbis.vrem.model.exhibition.Exhibition;
import ch.unibas.dmi.dbis.vrem.model.exhibition.Room;
import java.util.LinkedList;
import java.util.List;
import org.bson.BsonReader;
import org.bson.BsonType;
import org.bson.BsonWriter;
import org.bson.codecs.Codec;
import org.bson.codecs.DecoderContext;
import org.bson.codecs.EncoderContext;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.types.ObjectId;


public class ExhibitionCodec implements Codec<Exhibition> {

    public static final String FIELD_NAME_ID = "_id";
    public static final String FIELD_NAME_KEY = "key";
    public static final String FIELD_NAME_TEXT = "name";
    public static final String FIELD_NAME_DESCRIPTION = "description";
    public static final String FIELD_NAME_ROOMS = "rooms";

    private final Codec<Room> codec;

    public ExhibitionCodec(CodecRegistry registry) {
        this.codec = registry.get(Room.class);
    }

    @Override
    public Exhibition decode(BsonReader reader, DecoderContext decoderContext) {
        reader.readStartDocument();
        ObjectId id = null;
        String name = null;
        String description = null;
        String key = null;
        List<Room> rooms = new LinkedList<>();

        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
            switch (reader.readName()) {
                case FIELD_NAME_ID:
                    id = reader.readObjectId();
                    break;
                case FIELD_NAME_TEXT:
                    name = reader.readString();
                    break;
                case FIELD_NAME_DESCRIPTION:
                    description = reader.readString();
                    break;
                case FIELD_NAME_KEY:
                    key = reader.readString();
                    break;
                case FIELD_NAME_ROOMS:
                    reader.readStartArray();
                    while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
                        rooms.add(this.codec.decode(reader, decoderContext));
                    }
                    reader.readEndArray();
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.readEndDocument();
        final Exhibition exhibition = new Exhibition(id, key, name, description);
        for (Room room : rooms) {
            exhibition.addRoom(room);
        }
        return exhibition;
    }

    @Override
    public void encode(BsonWriter writer, Exhibition value, EncoderContext encoderContext) {
        writer.writeStartDocument();
        writer.writeObjectId(FIELD_NAME_ID, value.id);
        writer.writeString(FIELD_NAME_KEY, value.key);
        writer.writeString(FIELD_NAME_TEXT, value.name);
        writer.writeString(FIELD_NAME_DESCRIPTION, value.description);
        writer.writeName(FIELD_NAME_ROOMS);
        writer.writeStartArray();
        for (Room room : value.getRooms()) {
            this.codec.encode(writer, room, encoderContext);
        }
        writer.writeEndArray();
        writer.writeEndDocument();
    }

    @Override
    public Class<Exhibition> getEncoderClass() {
        return Exhibition.class;
    }
}

File: src/main/kotlin/ch/unibas/dmi/dbis/vrem/kotlin/database/codec/ExhibitionCodec.kt
package ch.unibas.dmi.dbis.vrem.kotlin.database.codec

import ch.unibas.dmi.dbis.vrem.database.codec.ExhibitionCodec
import ch.unibas.dmi.dbis.vrem.kotlin.model.exhibition.Exhibition
import ch.unibas.dmi.dbis.vrem.kotlin.model.exhibition.Room
import org.bson.BsonReader
import org.bson.BsonType
import org.bson.BsonWriter
import org.bson.codecs.Codec
import org.bson.codecs.DecoderContext
import org.bson.codecs.EncoderContext
import org.bson.codecs.configuration.CodecRegistry
import org.bson.types.ObjectId
import org.litote.kmongo.id.toId

/**
 * TODO: Write JavaDoc
 * @author loris.sauter
 */
class ExhibitionCodec(registry: CodecRegistry): Codec<Exhibition> {

    companion object{
        const val FIELD_NAME_ID = "_id"
        const val FIELD_NAME_NAME = "name"
        const val FIELD_NAME_DESCRIPTION = "description"
        const val FIELD_NAME_ROOMS = "rooms"
    }

    private val codec = registry.get(Room::class.java)

    override fun getEncoderClass(): Class<Exhibition> {
        return Exhibition::class.java
    }

    override fun encode(writer: BsonWriter?, value: Exhibition?, encoderContext: EncoderContext?) {
        writer!!.writeStartDocument()
        writer.writeObjectId(ExhibitionCodec.FIELD_NAME_ID, ObjectId(value!!.id.toString()))
        writer.writeString(ExhibitionCodec.FIELD_NAME_NAME, value!!.name)
        writer.writeString(ExhibitionCodec.FIELD_NAME_DESCRIPTION, value.description)
        writer.writeName(ExhibitionCodec.FIELD_NAME_ROOMS)
        writer.writeStartArray()
        for (room in value.rooms) {
            codec.encode(writer, room, encoderContext)
        }
        writer.writeEndArray()
        writer.writeEndDocument()
    }

    override fun decode(reader: BsonReader?, decoderContext: DecoderContext?): Exhibition {
        reader!!.readStartDocument()
        var id: ObjectId? = null
        var name: String? = null
        var description: String? = null
        val rooms: MutableList<Room> = mutableListOf()

        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
            when (reader.readName()) {
                ExhibitionCodec.FIELD_NAME_ID -> id = reader.readObjectId()
                ExhibitionCodec.FIELD_NAME_NAME -> name = reader.readString()
                ExhibitionCodec.FIELD_NAME_DESCRIPTION -> description = reader.readString()
                ExhibitionCodec.FIELD_NAME_ROOMS -> {
                    reader.readStartArray()
                    while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
                        rooms.add(codec.decode(reader, decoderContext))
                    }
                    reader.readEndArray()
                }
                else -> reader.skipValue()
            }
        }
        reader.readEndDocument()
        val exhibition = Exhibition(id!!.toId(),name!!, description!!)
        for (room in rooms) {
            exhibition.addRoom(room)
        }
        return exhibition
    }
}